src/Avalonia.Controls/TabControl.cs

```diff
@@ -83,67 +100,82 @@ public Dock TabStripPlacement
             set { SetValue(TabStripPlacementProperty, value); }
         }
 
-        protected override IItemContainerGenerator CreateItemContainerGenerator()
+        /// <summary>
+        /// Gets or sets the data template used to display the content of the control.
+        /// </summary>
+        public IDataTemplate ContentTemplate
         {
-            // TabControl doesn't actually create items - instead its TabStrip and Carousel
-            // children create the items. However we want it to be a SelectingItemsControl
-            // so that it has the Items/SelectedItem etc properties. In this case, we can
-            // return a null ItemContainerGenerator to disable the creation of item containers.
-            return null;
+            get { return GetValue(ContentTemplateProperty); }
+            set { SetValue(ContentTemplateProperty, value); }
         }
 
-        protected override void OnTemplateApplied(TemplateAppliedEventArgs e)
+        /// <summary>
+        /// Gets or sets the currently selected content.
+        /// </summary>
+        /// <value>
+        /// The content of the selected.
+        /// </value>
+        public object SelectedContent
         {
-            base.OnTemplateApplied(e);
-
-            TabStrip = e.NameScope.Find<IControl>("PART_TabStrip");
-            Pages = e.NameScope.Find<IControl>("PART_Content");
+            get { return GetValue(SelectedContentProperty); }
+            set { SetValue(SelectedContentProperty, value); }
         }
 
         /// <summary>
-        /// Selects the content of a tab item.
+        /// Gets or sets the template for the currently selected content.
         /// </summary>
-        /// <param name="o">The tab item.</param>
-        /// <returns>The content.</returns>
-        private static object SelectContent(object o)
+        /// <value>
+        /// The selected content template.
+        /// </value>
+        public IDataTemplate SelectedContentTemplate
+        {
+            get { return GetValue(SelectedContentTemplateProperty); }
+            set { SetValue(SelectedContentTemplateProperty, value); }
+        }
+
+        protected override IItemContainerGenerator CreateItemContainerGenerator()
+        {
+            return new TabItemContainerGenerator(this);
+        }
+
+        protected override void OnTemplateApplied(TemplateAppliedEventArgs e)
         {
-            var content = o as IContentControl;
+            base.OnTemplateApplied(e);
+
+            ItemsPresenterPart = e.NameScope.Find<ItemsPresenter>("PART_ItemsPresenter");
 
-            if (content != null)
+            if (ItemsPresenterPart == null)
             {
-                return content.Content;
+                throw new NotSupportedException("ItemsPresenter not found.");
```

Comment:
You can use the [`INameScope.Get`](https://github.com/AvaloniaUI/Avalonia/blob/dceabb95d99baacc431df7a0a195c905699fc1c1/src/Avalonia.Styling/Controls/NameScopeExtensions.cs#L48) extension method to throw an exception if the requested part isn't found.